### PR TITLE
todo_app

### PR DESCRIPTION
**Pull request by boriskerekesha**

### CODE REVIEW

**Comment by boriskerekesha** [comment]
https://boriskerekesha.github.io/react_todo-app/

File: src/App.tsx [hunk 1]
```diff
@@ -1,93 +1,61 @@
 /* eslint-disable jsx-a11y/control-has-associated-label */
-import React from 'react';
+import React, { useEffect, useState } from 'react';
+import { Routes, Route } from 'react-router-dom';
+import Footer from './components/Footer';
+import NewTodoField from './components/NewTodoField';
+import TodoList from './components/TodoList';
+import { ToggleAllButton } from './components/ToggleAllButton/ToggleAllButton';
```
Comment (TarasHoliuk) [inline src/App.tsx:7]:
```suggestion
import { ToggleAllButton } from './components/ToggleAllButton';
```
you have index file there

File: src/components/NewTodoField/NewTodoField.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  useEffect(() => {
+    if (newTitleInput.current) {
+      newTitleInput.current.focus();
+    }
+  });
+
+  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+
+    if (!value) {
```
Comment (TarasHoliuk) [inline src/components/NewTodoField/NewTodoField.tsx:23]: trim value. because now I can add todo with one space `' '`

File: src/components/TodoItem/TodoItem.tsx [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    if (titleInput.current) {
+      titleInput.current.focus();
+    }
+  });
+
+  const setId = () => `toggle-${completed ? 'completed' : 'view'}-${id}`;
+  const removeItem = (itemId: number) => setItems(currentItems => currentItems
+    .filter(itemTodo => itemTodo.id !== itemId));
+
+  const upDateItem = (text: string) => setItems(currentItem => {
```
Comment (TarasHoliuk) [inline src/components/TodoItem/TodoItem.tsx:28]:
```suggestion
  const updateItem = (text: string) => setItems(currentItems => {
```

File: src/components/TodoItem/TodoItem.tsx [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    item.title = text;
```
Comment (TarasHoliuk) [inline src/components/TodoItem/TodoItem.tsx:29]: it seems like a dirty hack ;)
you call `setItems` just to trigger re-render
better pass todo.id to this function, and map current items (change item with the same id)

File: src/components/TodoList/TodoList.tsx [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        return todo.completed;
+      default:
+        return true;
+    }
+  });
+
+  const visibleItems = filter ? getVisibleItems(items, filter) : [...items];
+
+  return (
+    <ul className="todo-list" data-cy="todoList">
```
Comment (TarasHoliuk) [inline src/components/TodoList/TodoList.tsx:33]:
```suggestion
    <ul className="todo-list" data-cy="todosList">
```
that's not your mistake, it is mismatch between initial HTML and tests, but it should fix tests :arrow_up:

File: src/components/TodoItem/TodoItem.tsx [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  const [isEditing, setIsEditing] = useState(false);
+  const titleInput = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (titleInput.current) {
+      titleInput.current.focus();
+    }
+  });
+
+  const setId = () => `toggle-${completed ? 'completed' : 'view'}-${id}`;
```
Comment (TarasHoliuk) [inline src/components/TodoItem/TodoItem.tsx:24]: That's nice, but why did you add it (common id would be enough)?

File: src/components/ToggleAllButton/ToggleAllButton.tsx [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  const isChecked = !items.find(todo => todo.completed === false);
+
+  const setStatus = (status: boolean) => {
+    setItems(currentTodo => {
+      currentTodo.forEach(todo => {
+        if (todo.completed === status) {
+          return;
+        }
+
+        todo.completed = status;
```
Comment (TarasHoliuk) [inline src/components/ToggleAllButton/ToggleAllButton.tsx:20]: the same as above (just map todos)

**Review by TarasHoliuk — CHANGES_REQUESTED** [review]
GJ! :+1: 
The app works fast :fire: 
But I've added a few comments to improve your code

File: src/components/TodoItem/TodoItem.tsx [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+  useEffect(() => {
+    if (titleInput.current) {
+      titleInput.current.focus();
+    }
+  });
+
+  const removeItem = (itemId: number) => setItems(currentItems => currentItems
+    .filter(itemTodo => itemTodo.id !== itemId));
+
+  const upDateItem = (itemId: number) => setItems(currentItems => currentItems
```
Comment (vlad-demchuk) [inline src/components/TodoItem/TodoItem.tsx:27]:
```suggestion
  const updateItem = (itemId: number) => setItems(currentItems => currentItems
```

**Review by vlad-demchuk — APPROVED** [review]
GJ!
